doc: applications: nrf_desktop: Add HID report providers doc (#23838)
nrfconnect/sdk-nrf: applications/nrf_desktop/doc/hid_provider_keyboard.rst

If there is no space in the queue to enqueue a new key state change, the oldest element is released.

Report expiration
-----------------

With the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_KEYBOARD_KEYPRESS_EXPIRATION <config_desktop_app_options>` Kconfig option, you can set the amount of time after which a queued key will be considered expired.
For backwards compatibility, the default value for this option can be set using the deprecated :ref:`CONFIG_DESKTOP_HID_REPORT_EXPIRATION <config_desktop_app_options>` Kconfig option.
The higher the value, the longer the period from which the nRF Desktop application will recall pressed keys when the connection with HID host is established.

Handling keys state
===================

The module selects the :ref:`CONFIG_DESKTOP_KEYS_STATE <config_desktop_app_options>` Kconfig option to enable the :ref:`nrf_desktop_keys_state`.
The utility is used to track the state of active keys after the connection with the HID host is established.

Implementation details
**********************

> On initialization, the module announces its presence by sending :c:struct:`hid_report_provider_event` event with appropriate report ID and implementation of callbacks from HID report provider API (:c:struct:`hid_report_provider_api`).

Reviewer: ```suggestion
On initialization, the module announces its presence by sending a :c:struct:`hid_report_provider_event` event with the appropriate report ID and implementation of callbacks from the HID report provider API (:c:struct:`hid_report_provider_api`).
```

Separate :c:struct:`hid_report_provider_event` events are sent for each report ID.

> The module supports keyboard reports from Report protocol and Boot protocol which can be enabled using :ref:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD <config_desktop_app_options>`.

Reviewer: ```suggestion
The module supports keyboard reports from the Report protocol and the Boot protocol, which you can enable using the :ref:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD <config_desktop_app_options>` Kconfig option.
```

The :ref:`nrf_desktop_hid_state` receives the events and fills them with its own implementation of callbacks from the HID state API (:c:struct:`hid_state_api`).

> After that process the modules can communicate by calling each other callbacks.

Reviewer: ```suggestion
After that process, the modules can communicate by callbacks.
```

> The module also subscribes to :c:struct:`button_event` events to get information about the button presses.

Reviewer: ```suggestion
The module also subscribes to the :c:struct:`button_event` event to get information about the button presses.
```

> The module sends :c:struct:`hid_report_event` events to appropriate subscriber when it is requested by the :ref:`nrf_desktop_hid_state`.

Reviewer: ```suggestion
The module sends :c:struct:`hid_report_event` events to an appropriate subscriber when it is requested by the :ref:`nrf_desktop_hid_state`.
```

Linking input data with the right HID report
============================================

Out of all available input data types, the module collects button events.
The button events are stored in the :c:struct:`report_data` structure.
The ``button_event`` is the source of this type of data.

To indicate a change to this input data, the module overwrites the value that is already stored.

> Since keys on the board can be associated to a HID usage ID, and thus be part of different HID reports, the first step is to identify if the key belongs to a HID report that is provided by this module.

Reviewer: ```suggestion
Since keys on the board can be associated with a HID usage ID and thus be part of different HID reports, the first step is to identify if the key belongs to a HID report that is provided by this module.
```

This is done by obtaining the key mapping from the :ref:`nrf_desktop_hid_keymap`.

Once the mapping is obtained, the application checks if the report to which the usage belongs is connected:

* If the report is connected and the :ref:`nrf_desktop_hid_eventq` instance is empty, the module stores the report and calls the ``trigger_report_send`` callback from the :c:struct:`hid_state_api` to notify |HID state| about the new data.
* If the report is not connected or the :ref:`nrf_desktop_hid_eventq` instance is not empty, the value is enqueued in the :ref:`nrf_desktop_hid_eventq` instance.

> The difference between these operations is that storing value onto the queue (second case) preserves the order of input events.

Reviewer: ```suggestion
The difference between these operations is that storing a value onto the queue (second case) preserves the order of input events.
```

See the following section for more information about storing data before the connection.

Storing input data before the connection
========================================

The button data is stored before the connection.

The reason for this operation is to allow to track key presses that happen right after the device is woken up, but before it is able to connect to the HID host.

> When the device is disconnected and the input event with the button data is received, the data is stored onto the :ref:`nrf_desktop_hid_eventq` instance, a member of :c:struct:`report_data` structure.
> This queue preserves an order at which input data events are received.

Reviewer: ```suggestion
When the device is disconnected and the input event with the button data is received, the data is stored onto the :ref:`nrf_desktop_hid_eventq` instance, a member of the :c:struct:`report_data` structure.
This queue preserves an order in which input data events are received.
```

Storing limitations
-------------------

> The number of events that can be inserted into the queue is limited using the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_KEYBOARD_EVENT_QUEUE_SIZE <config_desktop_app_options>` Kconfig option.

Reviewer: ```suggestion
You can limit the number of events that be inserted into the queue using the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_KEYBOARD_EVENT_QUEUE_SIZE <config_desktop_app_options>` Kconfig option.
```

Discarding events
------------------

When there is no space for a new input event, the module tries to free space by discarding the oldest event in the queue.

> Events stored in the queue are automatically discarded after the period defined by :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_KEYBOARD_KEYPRESS_EXPIRATION <config_desktop_app_options>` option.

Reviewer: ```suggestion
Events stored in the queue are automatically discarded after the period defined by the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_KEYBOARD_KEYPRESS_EXPIRATION <config_desktop_app_options>` option.
```

>     When discarding an event from the queue, the module checks if the key associated with the event is pressed.
>     This is to avoid missing key releases for earlier key presses when the keys from the queue are replayed to the host.
>     If a key release is missed, the host could stay with a key that is permanently pressed.
>     The discarding mechanism ensures that the host will always receive the correct key sequence.
> 
>     .. note::
>         The module can only discard an event if the event does not overlap any button that was pressed but not released, or if the button itself is pressed.
>         The event is released only when the following conditions are met:
> 
>         * The associated key is not pressed anymore.
>         * Every key that was pressed after the associated key had been pressed is also released.

Reviewer: ```suggestion
When discarding an event from the queue, the module checks if the key associated with the event is pressed.
This is to avoid missing key releases for earlier key presses when the keys from the queue are replayed to the host.
If a key release is missed, the host could stay with a key that is permanently pressed.
The discarding mechanism ensures that the host will always receive the correct key sequence.

.. note::
   The module can only discard an event if the event does not overlap any button that was pressed but not released, or if the button itself is pressed.
   The event is released only when the following conditions are met:

   * The associated key is not pressed anymore.
   * Every key that was pressed after the associated key had been pressed is also released.
```

> If there is no space to store the input event in the queue and no old event can be discarded, the entire content of the queue is dropped to ensure the sanity.
> 
> Once connection is established, the elements of the queue are replayed one after the other to the host, in a sequence of consecutive HID reports.

Reviewer: ```suggestion
If there is no space to store the input event in the queue and no old event can be discarded, the entire content of the queue is dropped to ensure sanity.

Once the connection is established, the elements of the queue are replayed one after the other to the host, in a sequence of consecutive HID reports.
```
